File: gdownload.py
#!/usr/bin/python
# -*- coding: utf-8 -*- 

import sys
import os
import re
from lxml.html import parse
from urllib import urlopen, URLopener, urlretrieve

"""
    - Downloads any project (sub)folder from github, to avoid downloading tests, demos, readme, etc, etc
    - Includes all the tree structure (and files) from the selected folder
    - It defaults to php-maxcdn library (/MaxCDN/php-maxcdn) and target folder defaults to ./ (current dir)
    - Download happens in current local directory by default 
    - Example, download only the /src folder of MaxCDN/php-maxcdn:
    $ cd my-project-folder/libraries/php-maxcdn/
    $ python updatelib.py "/MaxCDN/php-maxcdn/tree/master/src" /my/target/dir/
    - Tip: copy the full github folder path from the address bar, without the https://github.com part
    
"""
class getSrcFromGit():
    
    github = "https://github.com"
    githubraw = "https://raw.githubusercontent.com"
    localfolder = ""
    
    def __init__(self, args):
        
        if len(args) == 1:
            print "Github project required!"
            print "Usage: $ python gdownload.py \"/githubuser/project/tree/master/some/folder\" /my/target/dir/"
            exit()
        
        if len(args) > 1: 
            self.library = args[1]
        
        if len(args) > 2:
            self.target_dir = args[2]
        else: 
            self.target_dir = "./"

        if not os.path.exists(self.target_dir):
            os.makedirs(self.target_dir)
        
        self.url = self.github + self.library
        
        try:
            matched = re.search(r'(\/)[a-zA-Z0-9\-\_\.]+(\/)', self.library, re.M|re.I)
            print "Loading %s source from github... (%s)" % (matched.group().replace('/',''), self.url)
            parse(urlopen(self.url)).getroot()
        except IOError:
            print "Invalid github URL, please check"
            exit()
    
        exit()


    def load_source(self):
        git = parse(urlopen(self.url)).getroot()
        return git

    def process_source(self, url=None):

        if url is not None:
            self.url = url

        source = self.load_source()

        # This comes from github.com site, it could change in the future
        body = source.cssselect(".css-truncate .js-navigation-open")

        dir_list = []

        for f in body:
            
            current_item = f.text_content().split('.')

            # Quick and dirty way to see if is a file or folder
            if len(current_item) > 1:
                print 'Downloading %s to %s' % (f.text_content(), (self.target_dir + self.localfolder + f.text_content()))
                urlretrieve (self.githubraw + self.library.replace('/tree', '') + '/' + self.localfolder + f.text_content(), self.target_dir + self.localfolder + f.text_content())
            else:
                dir_list.append(f.text_content())

        for item in sorted(dir_list):

            self.localfolder = self.localfolder + item + '/'
            print "Creating subfolder: %s" % (self.target_dir + self.localfolder)

            if not os.path.exists(self.target_dir + self.localfolder):
                os.makedirs(self.target_dir + self.localfolder)

            self.process_source(self.url + '/' + item)


if __name__ == '__main__':
    getSrcFromGit(sys.argv).process_source()
    print "Download complete!"

File: gdownload.php
<?php

/**
 * Package LibUpdater
 * Updates github libraries
 */
Class LibUpdater {
    
    public $lib;
    public $path;
    public $libpath;
    public $html;
    public $urls;
    public $localfolder = '';
    public $targetfolder = './tests';
    public $foldertree = [];

    const GITHUB_URL = 'https://github.com';
    const GITHUB_RAW_URL = 'https://raw.githubusercontent.com';

    public function __construct( ...$params )
    {

        // /MaxCDN/php-maxcdn/tree/master/src

        if ( count( $params ) == 0 ){
            echo "Github project required!" . PHP_EOL;
            echo "Usage: $ php gdownload.php \"/githubuser/project/tree/master/some/folder\" /my/target/dir/" . PHP_EOL;
            exit;
        }

        if ( count( $params ) > 0 ) {
            $libpath = $params[0];
        }

        if ( count( $params ) > 1 ) {
            $this->targetfolder = $params[1];
        }

        if ( !is_dir( $this->targetfolder ) ) {
            echo 'Target folder is: [' . $this->targetfolder . ']' . PHP_EOL;
            mkdir( $this->targetfolder, 0775, true );
        } else {
            echo 'Target folder [ '. $this->targetfolder . ']  already exists...' . PHP_EOL;
        }

        // Allow HTML5
        libxml_use_internal_errors(true);
        $this->libpath = $libpath;
        $this->path = $this->libpath;
        $this->url = self::GITHUB_URL . $this->libpath;
        echo 'Loading [' . substr( $this->path, 1, strpos( $this->path, '/', 2 ) - 1 ) . '] from github... (' . self::GITHUB_URL . $this->path . ')' . PHP_EOL;

    }
    
    public function parseHTML( $url = null )
    {

        $url = !empty( $url ) ? $url : $this->url;
        $this->lib = file_get_contents( $url );

        $doc = new DOMDocument();
        $doc->loadHTML( $this->lib );

        $this->html = simplexml_import_dom( $doc );

        $container = $this->html->body->xpath( '//*[starts-with(@class, "css-truncate")]' );

        $this->urls = $container[0]->xpath( '//*[starts-with(@class, "js-navigation-open")]' );

        return $this->urls;

    }
    
    public function buildFolders( $root_folder )
    {
        
        $this->foldertree['remote'][] = $this->url;
        $this->foldertree['local'][] = $this->targetfolder . $this->localfolder;
        
        foreach( $root_folder as $key => $folder ) {
            
            if ( ( strpos( $folder[0], '.' ) !== false ) || $folder[0] == '..' ) {
                continue;
            }
            
            if ( $folder['title'] == 'This path skips through empty directories' ) {
                $folder_clean = trim( str_replace( $this->libpath, '', $folder['href'] ), '/' );
                $subfolder = explode( '/', $folder_clean );
            } else {
                $subfolder = (array) $folder['title'][0];
            }
            
            $this->foldertree['remote'][] = self::GITHUB_URL . $folder['href'];
            
            $this->localfolder = str_replace( $this->libpath, '', $folder['href'] );
            
            $this->foldertree['local'][] = $this->targetfolder . $this->localfolder;
            
            if ( !is_dir( $this->targetfolder . $this->localfolder ) ) {
                echo 'Creating dir ' . $this->targetfolder . $this->localfolder . PHP_EOL;
                if ( !mkdir( $this->targetfolder . $this->localfolder, 0775, true ) ) {
                    echo 'Can\'t mkdir: ' . $this->targetfolder . $this->localfolder . PHP_EOL;
                };
            } else {
                echo 'Directory ' . $this->targetfolder . $this->localfolder . ' already exists, skipping' . PHP_EOL;
            }
            
            $this->buildFolders( $this->parseHTML( self::GITHUB_URL . $folder['href'] ) );
            
        }
        
    }
    
    public function downloadFiles()
    {
        
        foreach( $this->foldertree['remote'] as $key => $folder ) {
            
            $subfolders = $this->parseHTML( $folder );
            
            foreach( $subfolders as $sub ) {
                
                
                if ( $sub != '..' ) {
                    if( strpos( $sub, '.' ) !== false ) {
                        
                        // Fix URL 1
                        if ( strpos( $this->libpath, '/master' ) === false ) {
                            $this->libpath .= '/master';
                        }

                        // Fix URL 2
                        $this->foldertree['remote'][$key] = str_replace( self::GITHUB_URL, self::GITHUB_RAW_URL, $this->foldertree['remote'][$key] );
                        $this->foldertree['remote'][$key] = str_replace( [ 'blob/', 'tree/' ], '', $this->foldertree['remote'][$key] );
                        
                        // Download file
                        echo 'Downloading ' . $this->foldertree['remote'][$key] . '/' . $sub . ' to ' . $this->foldertree['local'][$key] . '/' . $sub . PHP_EOL;
                        $file_contents = file_get_contents( $this->foldertree['remote'][$key] . '/' . $sub );
                        file_put_contents( $this->foldertree['local'][$key] . '/' . $sub, $file_contents );
                        
                    } 
                } 
                
            }
            
        }
        
        return true;
    }

}